### PR TITLE
Deprecate pc.events in favor of inheritance based pc.EventHandler

### PR DESCRIPTION
**Pull request by willeastcott**
**PROBLEM**
Currently, JSDoc cannot enumerate the on/off/once/fire/hasEvent functions on many PlayCanvas 'classes'. The only workaround would be to copy/paste JSDoc comments for those functions onto every class that calls pc.events.attach, which would be a lot of duplication.

Also, the TypeScript bindings do not have definitions for these functions.

Furthermore, the architecture of the pc.events namespace does not conform with the rest of the engine, which is class based and relies on inheritance.

**SOLUTION**
This PR deprecates pc.events and introduces pc.EventHandler, an abstract base class that other classes can inherit from to gain event handling capabilities. For now, the old behavior/API is still available, it's just internal/private.

There are three remaining uses of pc.events.attach that are internal to the engine: 

1. pc.ComponentSystem
2. pc.script
3. Legacy script instances

Ideally, these would be refactored to do something different, but for now, they continue to use the now private API. If that is done, a deprecation warning message can be added to pc.events.attach.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/core/event-handler.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            } else {
+                if (this._callbackActive[name] === this._callbacks[name])
+                    this._callbackActive[name] = this._callbackActive[name].slice();
+
+                callbacks = this._callbacks[name].slice();
+            }
+
+            // TODO: What does callbacks do here?
+            // In particular this condition check looks wrong: (i < (callbacks || this._callbackActive[name]).length)
+            // Because callbacks is not an integer
```
Comment (c0debon) [inline src/core/event-handler.js]: integer->array?

Comment (willeastcott) [reply src/core/event-handler.js]: I'm not sure! I've merely refactored the code in events.js to a base class format. The function contents, including the comments, are identical to what they were before.

File: src/framework/stats.js [hunk 2]
```diff
@@ -86,6 +86,4 @@ pc.ApplicationStats = function (device) {
             return pc.Application._currentApplication.batcher._stats;
         }
     });
-
-    pc.events.attach(this);
```
Comment (vkalpias) [inline src/framework/stats.js]: We don't seem to add a base class in this case is this intentional?

Comment (willeastcott) [reply src/framework/stats.js]: Yes. I can find no evidence that pc.ApplicationStats does anything with events, so maybe this was a copy and paste mistake when the constructor was originally created.

**Pull request merged**